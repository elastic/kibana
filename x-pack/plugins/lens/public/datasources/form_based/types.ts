/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { DragDropIdentifier } from '@kbn/dom-drag-drop';
import { AggregateQuery } from '@kbn/es-query';
import { Datatable } from '@kbn/expressions-plugin/common';
import { VisualizeFieldContext } from '@kbn/ui-actions-plugin/public';
import { IndexPatternRef, TextBasedLayerColumn } from './esql_layer/types';
import type { IndexPattern, IndexPatternField, DragDropOperation } from '../../types';
import type { IncompleteColumn, GenericIndexPatternColumn } from './operations';
import { VisualizeEditorContext } from '../../types';
import { ValueFormatConfig } from './operations/definitions/column_types';

export type {
  GenericIndexPatternColumn,
  OperationType,
  IncompleteColumn,
  FieldBasedIndexPatternColumn,
  FiltersIndexPatternColumn,
  RangeIndexPatternColumn,
  TermsIndexPatternColumn,
  DateHistogramIndexPatternColumn,
  MinIndexPatternColumn,
  MaxIndexPatternColumn,
  AvgIndexPatternColumn,
  CardinalityIndexPatternColumn,
  SumIndexPatternColumn,
  MedianIndexPatternColumn,
  StandardDeviationIndexPatternColumn,
  PercentileIndexPatternColumn,
  PercentileRanksIndexPatternColumn,
  CountIndexPatternColumn,
  LastValueIndexPatternColumn,
  CumulativeSumIndexPatternColumn,
  CounterRateIndexPatternColumn,
  DerivativeIndexPatternColumn,
  MovingAverageIndexPatternColumn,
  FormulaIndexPatternColumn,
  MathIndexPatternColumn,
  OverallSumIndexPatternColumn,
  StaticValueIndexPatternColumn,
  TimeScaleIndexPatternColumn,
} from './operations';

export type { FormulaPublicApi } from './operations/definitions/formula/formula_public_api';

export type DraggedField = DragDropIdentifier & {
  field: IndexPatternField;
  indexPatternId: string;
};

export interface TextBasedLayer {
  type: 'esql';
  index?: string;
  indexPatternId?: string;
  query?: AggregateQuery | undefined;
  table?: Datatable;
  columns: TextBasedLayerColumn[];
  timeField?: string;
  params?: {
    format?: ValueFormatConfig;
  };
  errors?: Error[];
}

export interface FormBasedLayer {
  type: 'form' | undefined;
  columnOrder: string[];
  columns: Record<string, GenericIndexPatternColumn>;
  // Each layer is tied to the index pattern that created it
  indexPatternId: string;
  linkToLayers?: string[];
  // Partial columns represent the temporary invalid states
  incompleteColumns?: Record<string, IncompleteColumn | undefined>;
  sampling?: number;
  ignoreGlobalFilters?: boolean;
}

export interface FormBasedPersistedState {
  layers: Record<string, Omit<FormBasedLayer, 'indexPatternId'> | TextBasedLayer>;
}

export type PersistedIndexPatternLayer = Omit<FormBasedLayer, 'indexPatternId'>;

export interface FormBasedPrivateState {
  currentIndexPatternId: string;
  indexPatternRefs: IndexPatternRef[];
  initialContext?: VisualizeFieldContext | VisualizeEditorContext;
  layers: Record<string, FormBasedLayer | TextBasedLayer>;
}

export interface DataViewDragDropOperation extends DragDropOperation {
  dataView: IndexPattern;
  column?: GenericIndexPatternColumn;
}
