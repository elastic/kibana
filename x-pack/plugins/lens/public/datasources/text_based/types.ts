/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { DatatableColumn } from '@kbn/expressions-plugin/public';
import type { AggregateQuery } from '@kbn/es-query';
import type { VisualizeFieldContext } from '@kbn/ui-actions-plugin/public';
import type { VisualizeEditorContext } from '../../types';

export interface TextBasedLayerColumn {
  columnId: string;
  fieldName: string;
  meta?: DatatableColumn['meta'];
}

export interface TextBasedField {
  id: string;
  field: string;
}

export interface TextBasedLayer {
  index: string;
  query: AggregateQuery | undefined;
  columns: TextBasedLayerColumn[];
  allColumns: TextBasedLayerColumn[];
  timeField?: string;
  errors?: Error[];
}

export interface TextBasedPersistedState {
  layers: Record<string, TextBasedLayer>;
}

export type TextBasedPrivateState = TextBasedPersistedState & {
  indexPatternRefs: IndexPatternRef[];
  fieldList: DatatableColumn[];
  initialContext?: VisualizeFieldContext | VisualizeEditorContext;
};

export interface IndexPatternRef {
  id: string;
  title: string;
  timeField?: string;
  name?: string;
}
