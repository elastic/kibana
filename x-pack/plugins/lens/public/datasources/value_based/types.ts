/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { Datatable, DatatableColumn } from '@kbn/expressions-plugin/public';
import type { VisualizeFieldContext } from '@kbn/ui-actions-plugin/public';
import type { VisualizeEditorContext } from '../../types';

export interface ValueBasedLayerColumn {
  columnId: string;
  fieldName: string;
  meta?: DatatableColumn['meta'];
}

export interface ValueBasedField {
  id: string;
  field: string;
}

export interface ValueBasedLayer {
  table: Datatable;
  columns: ValueBasedLayerColumn[];
  errors?: Error[];
}

export interface ValueBasedPersistedState {
  layers: Record<string, ValueBasedLayer>;
  initialContext?: VisualizeFieldContext | VisualizeEditorContext;
  fieldList?: DatatableColumn[];
}

export type ValueBasedPrivateState = ValueBasedPersistedState;
