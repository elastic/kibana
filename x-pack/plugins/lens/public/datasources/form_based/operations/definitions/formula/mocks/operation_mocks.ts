/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import type {
  GenericOperationDefinition,
  OperationDefinition,
  OperationDefinitionMap,
} from '../..';
import type { BaseIndexPatternColumn, GenericIndexPatternColumn } from '../../column_types';
import { getFilter } from '../../helpers';

interface PartialColumnParams {
  kql?: string;
  lucene?: string;
  shift?: string;
  reducedTimeRange?: string;
}

export function createOperationDefinitionMock<
  T extends keyof OperationDefinitionMap<BaseIndexPatternColumn>
>(
  operation: string,
  column: T extends keyof OperationDefinitionMap<BaseIndexPatternColumn>
    ? Partial<OperationDefinition<BaseIndexPatternColumn, T>>
    : Partial<OperationDefinition<BaseIndexPatternColumn, 'field'>>,
  {
    label = operation,
    dataType = 'number',
    isBucketed = false,
    scale = 'ratio',
    timeScale,
  }: Partial<GenericIndexPatternColumn> = {}
): GenericOperationDefinition {
  const { input = 'field', getErrorMessage, buildColumn, ...params } = column;
  const sharedColumnParams = {
    label,
    dataType,
    operationType: operation,
    isBucketed,
    scale,
    timeScale,
  };
  const sharedDefinitionParams = {
    input,
    getDefaultLabel: jest.fn(),
    isTransferable: jest.fn(),
    displayName: label,
    type: operation,
    getErrorMessage: getErrorMessage ?? jest.fn(),
    toExpression: jest.fn(),
  };
  if (input === 'field') {
    return {
      buildColumn:
        buildColumn ??
        jest.fn(({ field }, columnParams: PartialColumnParams) => ({
          sourceField: field.name,
          filter: getFilter(undefined, columnParams),
          reducedTimeRange: columnParams.reducedTimeRange,
          ...sharedColumnParams,
        })),
      onFieldChange: jest.fn(),
      toEsAggsFn: jest.fn(),
      getPossibleOperationForField:
        (params as unknown as OperationDefinition<BaseIndexPatternColumn, 'field'>)
          ?.getPossibleOperationForField ??
        jest.fn((arg) => ({
          scale,
          dataType,
          isBucketed,
        })),
      ...sharedDefinitionParams,
      ...params,
    } as OperationDefinition<BaseIndexPatternColumn, 'field'>;
  }
  if (input === 'fullReference') {
    return {
      buildColumn:
        buildColumn ??
        jest.fn(({ referenceIds }, columnParams: PartialColumnParams) => ({
          references: referenceIds,
          filter: getFilter(undefined, columnParams),
          ...sharedColumnParams,
        })),
      onFieldChange: jest.fn(),
      toEsAggsFn: jest.fn(),
      getPossibleOperation: jest.fn(() => ({
        scale: 'ratio',
        dataType: 'number',
        isBucketed: false,
      })),
      requiredReferences: [
        {
          input: ['field', 'managedReference'],
          validateMetadata: jest.fn(),
        },
      ],
      selectionStyle: 'field',
      ...sharedDefinitionParams,
      ...params,
    } as OperationDefinition<BaseIndexPatternColumn, 'fullReference'>;
  }
  return {
    buildColumn:
      buildColumn ??
      jest.fn((_, columnsParams: PartialColumnParams) => ({
        references: [],
        filter: getFilter(undefined, columnsParams),
        ...sharedColumnParams,
      })),
    getPossibleOperation: jest.fn(),
    createCopy: jest.fn(),
    ...sharedDefinitionParams,
    ...params,
  } as OperationDefinition<BaseIndexPatternColumn, 'none' | 'managedReference'>;
}
