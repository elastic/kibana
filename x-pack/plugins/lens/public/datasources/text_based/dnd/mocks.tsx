/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { DatasourceDimensionDropHandlerProps } from '../../../types';
import { TextBasedPrivateState } from '../types';

export const column1 = {
  columnId: 'columnId1',
  fieldName: 'category',
  meta: {
    type: 'string',
  },
};
export const column2 = {
  columnId: 'columnId2',
  fieldName: 'day_of_week_i',
  meta: {
    type: 'number',
  },
};
export const column3 = {
  columnId: 'columnId3',
  fieldName: 'products.base_price',
  meta: {
    type: 'number',
  },
};

export const emptyDimensionTarget = {
  layerId: 'first',
  groupId: 'y',
  isNewColumn: true,
  indexPatternId: 'indexId',
  filterOperations: () => true,
  humanData: {
    groupLabel: 'Horizontal axis',
    layerNumber: 1,
    position: 2,
    label: 'Empty dimension',
    nextLabel: '',
    canDuplicate: true,
  },
  columnId: 'newId',
  id: 'newId',
};

export const numericDraggedColumn = {
  id: column1.columnId,
  layerId: 'first',
  columnId: column1.columnId,
  groupId: 'y',
  isMetricDimension: false,
  indexPatternId: 'indexId',
  humanData: {
    label: 'category',
    groupLabel: 'Horizontal axis',
    position: 1,
    layerNumber: 1,
    canSwap: false,
    canDuplicate: false,
    canCombine: false,
    nextLabel: '',
  },
};

export const notNumericDraggedField = {
  field: 'category',
  id: 'category',
  humanData: {
    label: 'category',
  },
};

export const numericDraggedField = {
  field: 'products.base_price',
  id: 'products.base_price',
  humanData: {
    label: 'products.base_price',
  },
};

export const fieldList = [
  {
    columnId: 'category',
    fieldName: 'category',
    meta: {
      type: 'string',
    },
  },
  {
    columnId: 'currency',
    fieldName: 'currency',
    meta: {
      type: 'string',
    },
  },
  {
    columnId: 'products.base_price',
    fieldName: 'products.base_price',
    meta: {
      type: 'number',
    },
  },
  {
    columnId: 'products.price',
    fieldName: 'products.price',
    meta: {
      type: 'number',
    },
  },
];

export const defaultProps = {
  setState: jest.fn(),
  state: {
    layers: {
      first: {
        index: 'indexId',
        query: {
          sql: 'SELECT * FROM "kibana_sample_data_ecommerce"',
        },
        columns: [column1, column2, column3],
        allColumns: [...fieldList, column1, column2, column3],
        errors: [],
      },
    },
    fieldList,
    indexPatternRefs: [],
  },
  source: numericDraggedColumn,
  target: {
    layerId: 'first',
    groupId: 'y',
    isNewColumn: true,
    isMetricDimension: true,
    indexPatternId: 'indexId',
    filterOperations: () => true,
    humanData: {
      groupLabel: 'Horizontal axis',
      layerNumber: 1,
      position: 2,
      label: 'Empty dimension',
      nextLabel: 'products.price',
      canDuplicate: true,
    },
    columnId: column3.columnId,
    id: column3.columnId,
  },
  targetLayerDimensionGroups: [
    {
      groupId: 'x',
      groupLabel: 'Vertical axis',
      accessors: [
        {
          columnId: 'columnId3',
        },
      ],
      supportsMoreColumns: false,
      dataTestSubj: 'lnsXY_xDimensionPanel',
    },
    {
      groupId: 'y',
      groupLabel: 'Horizontal axis',
      accessors: [
        {
          columnId: 'columnId1',
          triggerIconType: 'disabled',
        },
      ],
      isMetricDimension: true,
      supportsMoreColumns: true,
      requiredMinDimensionCount: 1,
      dataTestSubj: 'lnsXY_yDimensionPanel',
      enableDimensionEditor: true,
    },
    {
      groupId: 'breakdown',
      groupLabel: 'Breakdown',
      accessors: [
        {
          columnId: 'columnId2',
          triggerIconType: 'colorBy',
        },
      ],
      supportsMoreColumns: false,
      dataTestSubj: 'lnsXY_splitDimensionPanel',
      requiredMinDimensionCount: 0,
      enableDimensionEditor: true,
    },
  ],
  dropType: 'duplicate_compatible',
} as unknown as DatasourceDimensionDropHandlerProps<TextBasedPrivateState>;
