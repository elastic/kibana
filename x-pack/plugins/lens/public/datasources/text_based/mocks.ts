/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { DragContextState } from '../../drag_drop';

export function createMockedDragDropContext(): jest.Mocked<DragContextState> {
  return {
    dragging: undefined,
    setDragging: jest.fn(),
    activeDropTarget: undefined,
    setActiveDropTarget: jest.fn(),
    keyboardMode: false,
    setKeyboardMode: jest.fn(),
    setA11yMessage: jest.fn(),
    dropTargetsByOrder: undefined,
    registerDropTarget: jest.fn(),
  };
}
