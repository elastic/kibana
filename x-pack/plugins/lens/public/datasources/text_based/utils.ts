/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
import type { DataViewsPublicPluginStart } from '@kbn/data-views-plugin/public';
import type { DataPublicPluginStart } from '@kbn/data-plugin/public';
import type { ExpressionsStart } from '@kbn/expressions-plugin/public';

import { type AggregateQuery, getIndexPatternFromSQLQuery } from '@kbn/es-query';
import type { DatatableColumn } from '@kbn/expressions-plugin/public';
import { generateId } from '../../id_generator';
import { fetchDataFromAggregateQuery } from './fetch_data_from_aggregate_query';

import type { IndexPatternRef, TextBasedPrivateState, TextBasedLayerColumn } from './types';
import type { DataViewsState } from '../../state_management';

export async function loadIndexPatternRefs(
  indexPatternsService: DataViewsPublicPluginStart
): Promise<IndexPatternRef[]> {
  const indexPatterns = await indexPatternsService.getIdsWithTitle();

  const timefields = await Promise.all(
    indexPatterns.map((p) => indexPatternsService.get(p.id).then((pat) => pat.timeFieldName))
  );

  return indexPatterns
    .map((p, i) => ({ ...p, timeField: timefields[i] }))
    .sort((a, b) => {
      return a.title.localeCompare(b.title);
    });
}

export const getAllColumns = (
  existingColumns: TextBasedLayerColumn[],
  columnsFromQuery: DatatableColumn[]
) => {
  // filter out columns that do not exist on the query
  const columns = existingColumns.filter((c) => {
    const columnExists = columnsFromQuery?.some((f) => f.name === c?.fieldName);
    if (columnExists) return c;
  });
  const allCols = [
    ...columns,
    ...columnsFromQuery.map((c) => ({ columnId: c.id, fieldName: c.id, meta: c.meta })),
  ];
  const uniqueIds: string[] = [];

  return allCols.filter((col) => {
    const isDuplicate = uniqueIds.includes(col.columnId);

    if (!isDuplicate) {
      uniqueIds.push(col.columnId);

      return true;
    }

    return false;
  });
};

export async function getStateFromAggregateQuery(
  state: TextBasedPrivateState,
  query: AggregateQuery,
  dataViews: DataViewsPublicPluginStart,
  data: DataPublicPluginStart,
  expressions: ExpressionsStart,
  frameDataViews?: DataViewsState
) {
  let indexPatternRefs: IndexPatternRef[] = frameDataViews?.indexPatternRefs.length
    ? frameDataViews.indexPatternRefs
    : await loadIndexPatternRefs(dataViews);
  const errors: Error[] = [];
  const layerIds = Object.keys(state.layers);
  const context = state.initialContext;
  const newLayerId = layerIds.length > 0 ? layerIds[0] : generateId();
  // fetch the pattern from the query
  const indexPattern = getIndexPatternFromTextBasedQuery(query);
  // get the id of the dataview
  let dataViewId = indexPatternRefs.find((r) => r.title === indexPattern)?.id ?? '';
  let columnsFromQuery: DatatableColumn[] = [];
  let allColumns: TextBasedLayerColumn[] = [];
  let timeFieldName;
  try {
    const dataView = dataViewId
      ? await dataViews.get(dataViewId)
      : await dataViews.create({
          title: indexPattern,
        });
    if (!dataViewId && !dataView.isPersisted()) {
      if (dataView && dataView.id) {
        if (dataView.fields.getByName('@timestamp')?.type === 'date') {
          dataView.timeFieldName = '@timestamp';
        } else if (dataView.fields.getByType('date')?.length) {
          const dateFields = dataView.fields.getByType('date');
          dataView.timeFieldName = dateFields[0].name;
        }
        dataViewId = dataView?.id;
        indexPatternRefs = [
          ...indexPatternRefs,
          {
            id: dataView.id,
            title: dataView.name,
            timeField: dataView.timeFieldName,
          },
        ];
      }
    }
    timeFieldName = dataView.timeFieldName;
    const table = await fetchDataFromAggregateQuery(query, dataView, data, expressions);
    columnsFromQuery = table?.columns ?? [];
    allColumns = getAllColumns(state.layers[newLayerId].allColumns, columnsFromQuery);
  } catch (e) {
    errors.push(e);
  }

  const tempState = {
    layers: {
      [newLayerId]: {
        index: dataViewId,
        query,
        columns: state.layers[newLayerId].columns ?? [],
        allColumns,
        timeField: timeFieldName,
        errors,
      },
    },
  };

  return {
    ...tempState,
    fieldList: columnsFromQuery ?? [],
    indexPatternRefs,
    initialContext: context,
  };
}

export function getIndexPatternFromTextBasedQuery(query: AggregateQuery): string {
  let indexPattern = '';
  // sql queries
  if ('sql' in query) {
    indexPattern = getIndexPatternFromSQLQuery(query.sql);
  }
  // other textbased queries....

  return indexPattern;
}
