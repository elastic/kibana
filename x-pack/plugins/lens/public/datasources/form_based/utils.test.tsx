/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React from 'react';
import { shallow } from 'enzyme';
import { createDatatableUtilitiesMock } from '@kbn/data-plugin/common/mocks';
import { getPrecisionErrorWarningMessages, cloneLayer } from './utils';
import type { FormBasedPrivateState, GenericIndexPatternColumn } from './types';
import type { FramePublicAPI } from '../../types';
import type { DocLinksStart } from '@kbn/core/public';
import { EuiLink } from '@elastic/eui';
import { TermsIndexPatternColumn } from './operations';
import { mountWithIntl } from '@kbn/test-jest-helpers';
import { FormattedMessage } from '@kbn/i18n-react';
import { FormBasedLayer } from './types';

describe('indexpattern_datasource utils', () => {
  describe('getPrecisionErrorWarningMessages', () => {
    const datatableUtilitites = createDatatableUtilitiesMock();
    let state: FormBasedPrivateState;
    let framePublicAPI: FramePublicAPI;
    let docLinks: DocLinksStart;

    beforeEach(() => {
      state = {
        layers: {
          id: {
            indexPatternId: 'one',
            columns: {
              col1: {
                operationType: 'terms',
                params: {
                  orderBy: {
                    type: 'alphabetical',
                  },
                },
              } as TermsIndexPatternColumn,
            },
          },
        },
      } as unknown as FormBasedPrivateState;
      framePublicAPI = {
        activeData: {
          id: {
            columns: [
              {
                id: 'col1',
                meta: {
                  sourceParams: {
                    hasPrecisionError: false,
                  },
                },
              },
            ],
          },
        },
        dataViews: {
          indexPatterns: {
            one: {
              getFieldByName: (x: string) => ({ name: x, displayName: x }),
            },
          },
        },
      } as unknown as FramePublicAPI;

      docLinks = {
        links: {
          aggs: {
            terms_doc_count_error: 'http://terms_doc_count_error',
          },
        },
      } as DocLinksStart;
    });
    test('should not show precisionError if hasPrecisionError is false', () => {
      expect(
        getPrecisionErrorWarningMessages(
          datatableUtilitites,
          state,
          framePublicAPI,
          docLinks,
          () => {}
        )
      ).toHaveLength(0);
    });

    test('should not show precisionError if hasPrecisionError is not defined', () => {
      delete framePublicAPI.activeData!.id.columns[0].meta.sourceParams!.hasPrecisionError;

      expect(
        getPrecisionErrorWarningMessages(
          datatableUtilitites,
          state,
          framePublicAPI,
          docLinks,
          () => {}
        )
      ).toHaveLength(0);
    });

    describe('precision error warning with accuracy mode', () => {
      const enableAccuracyButtonSelector =
        'button[data-test-subj="lnsPrecisionWarningEnableAccuracy"]';

      test('should show accuracy mode prompt if currently disabled', async () => {
        framePublicAPI.activeData!.id.columns[0].meta.sourceParams!.hasPrecisionError = true;
        (state.layers.id.columns.col1 as TermsIndexPatternColumn).params.accuracyMode = false;

        const setStateMock = jest.fn();

        const warningMessages = getPrecisionErrorWarningMessages(
          datatableUtilitites,
          state,
          framePublicAPI,
          docLinks,
          setStateMock
        );

        expect(warningMessages).toHaveLength(1);

        expect({ ...warningMessages[0], longMessage: '' }).toMatchSnapshot();

        const instance = mountWithIntl(<div>{warningMessages[0].longMessage}</div>);

        const enableAccuracyButton = instance.find(enableAccuracyButtonSelector);

        expect(enableAccuracyButton.exists()).toBeTruthy();

        enableAccuracyButton.simulate('click');

        expect(setStateMock).toHaveBeenCalledTimes(1);
      });

      test('should other suggestions if accuracy mode already enabled', async () => {
        framePublicAPI.activeData!.id.columns[0].meta.sourceParams!.hasPrecisionError = true;
        (state.layers.id.columns.col1 as TermsIndexPatternColumn).params.accuracyMode = true;

        const warningMessages = getPrecisionErrorWarningMessages(
          datatableUtilitites,
          state,
          framePublicAPI,
          docLinks,
          () => {}
        );

        expect(warningMessages).toHaveLength(1);

        expect({ ...warningMessages[0], longMessage: '' }).toMatchSnapshot();

        const instance = shallow(<div>{warningMessages[0].longMessage}</div>);

        expect(instance.exists(enableAccuracyButtonSelector)).toBeFalsy();

        expect(instance.find(FormattedMessage).props().id).toBe(
          'xpack.lens.indexPattern.precisionErrorWarning.accuracyEnabled'
        );
      });
    });

    test('if has precision error and sorting is by count ascending, show fix action and switch to rare terms', () => {
      framePublicAPI.activeData!.id.columns[0].meta.sourceParams!.hasPrecisionError = true;
      state.layers.id.columnOrder = ['col1', 'col2'];
      state.layers.id.columns = {
        col1: {
          operationType: 'terms',
          sourceField: 'category',
          params: {
            orderBy: {
              type: 'column',
              columnId: 'col2',
            },
            orderDirection: 'asc',
          },
        } as unknown as GenericIndexPatternColumn,
        col2: {
          operationType: 'count',
        } as unknown as GenericIndexPatternColumn,
      };
      const setState = jest.fn();
      const warnings = getPrecisionErrorWarningMessages(
        datatableUtilitites,
        state,
        framePublicAPI,
        docLinks,
        setState
      );

      expect(warnings).toHaveLength(1);
      expect({ ...warnings[0], longMessage: '' }).toMatchSnapshot();
      const DummyComponent = () => <>{warnings[0].longMessage}</>;
      const warningUi = shallow(<DummyComponent />);
      warningUi.find(EuiLink).simulate('click');
      const stateSetter = setState.mock.calls[0][0];
      const newState = stateSetter(state);
      expect(newState.layers.id.columns.col1.label).toEqual('Rare values of category');
      expect(newState.layers.id.columns.col1.params.orderBy).toEqual({
        type: 'rare',
        maxDocCount: 1,
      });
    });
  });

  describe('cloneLayer', () => {
    test('should clone layer with renewing ids', () => {
      expect(
        cloneLayer(
          {
            a: {
              columns: {
                '899ee4b6-3147-4d45-94bf-ea9c02e55d28': {
                  params: {
                    orderBy: {
                      type: 'column',
                      columnId: 'ae62cfc8-faa5-4096-a30c-f92ac59922a0',
                    },
                    orderDirection: 'desc',
                  },
                },
                'ae62cfc8-faa5-4096-a30c-f92ac59922a0': {
                  params: {
                    emptyAsNull: true,
                  },
                },
              },
              columnOrder: [
                '899ee4b6-3147-4d45-94bf-ea9c02e55d28',
                'ae62cfc8-faa5-4096-a30c-f92ac59922a0',
              ],
              incompleteColumns: {},
              indexPatternId: 'ff959d40-b880-11e8-a6d9-e546fe2bba5f',
            },
          } as unknown as Record<string, FormBasedLayer>,
          'a',
          'b',
          (id) => id + 'C'
        )
      ).toMatchSnapshot();
    });
  });
});
