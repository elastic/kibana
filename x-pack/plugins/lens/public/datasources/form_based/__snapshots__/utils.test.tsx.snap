// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`indexpattern_datasource utils cloneLayer should clone layer with renewing ids 1`] = `
Object {
  "a": Object {
    "columnOrder": Array [
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28",
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0",
    ],
    "columns": Object {
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28": Object {
        "params": Object {
          "orderBy": Object {
            "columnId": "ae62cfc8-faa5-4096-a30c-f92ac59922a0",
            "type": "column",
          },
          "orderDirection": "desc",
        },
      },
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0": Object {
        "params": Object {
          "emptyAsNull": true,
        },
      },
    },
    "incompleteColumns": Object {},
    "indexPatternId": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
  },
  "b": Object {
    "columnOrder": Array [
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28C",
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0C",
    ],
    "columns": Object {
      "899ee4b6-3147-4d45-94bf-ea9c02e55d28C": Object {
        "params": Object {
          "orderBy": Object {
            "columnId": "ae62cfc8-faa5-4096-a30c-f92ac59922a0C",
            "type": "column",
          },
          "orderDirection": "desc",
        },
      },
      "ae62cfc8-faa5-4096-a30c-f92ac59922a0C": Object {
        "params": Object {
          "emptyAsNull": true,
        },
      },
    },
    "incompleteColumns": Object {},
    "indexPatternId": "ff959d40-b880-11e8-a6d9-e546fe2bba5f",
  },
}
`;

exports[`indexpattern_datasource utils getPrecisionErrorWarningMessages if has precision error and sorting is by count ascending, show fix action and switch to rare terms 1`] = `
Object {
  "displayLocations": Array [
    Object {
      "id": "toolbar",
    },
    Object {
      "dimensionId": "col1",
      "id": "dimensionButton",
    },
  ],
  "fixableInEditor": true,
  "longMessage": "",
  "severity": "warning",
  "shortMessage": "This may be approximate depending on how the data is indexed. For more precise results, sort by rarity.",
}
`;

exports[`indexpattern_datasource utils getPrecisionErrorWarningMessages precision error warning with accuracy mode should other suggestions if accuracy mode already enabled 1`] = `
Object {
  "displayLocations": Array [
    Object {
      "id": "toolbar",
    },
    Object {
      "dimensionId": "col1",
      "id": "dimensionButton",
    },
  ],
  "fixableInEditor": true,
  "longMessage": "",
  "severity": "warning",
  "shortMessage": "This might be an approximation. For more precise results, use Filters or increase the number of Top Values.",
}
`;

exports[`indexpattern_datasource utils getPrecisionErrorWarningMessages precision error warning with accuracy mode should show accuracy mode prompt if currently disabled 1`] = `
Object {
  "displayLocations": Array [
    Object {
      "id": "toolbar",
    },
    Object {
      "dimensionId": "col1",
      "id": "dimensionButton",
    },
  ],
  "fixableInEditor": true,
  "longMessage": "",
  "severity": "warning",
  "shortMessage": "This might be an approximation. For more precise results, you can enable accuracy mode, but it increases the load on the Elasticsearch cluster.",
}
`;
